### PR TITLE
Adds tenant management APIs to developer facing Auth instance.

### PR DESCRIPTION
**Pull request by bojeil-google**
This includes getTenant, deleteTenant, listTenants, createTenant and updateTenant.
This expects TenantServerResponse to be returned for createTenant and updateTenant.

Expected results have not been confirmed. A followup PR will add integration tests for the above.

### CODE REVIEW

File: src/auth/auth.ts [hunk 1]
```diff
@@ -778,4 +778,95 @@ export class Auth extends BaseAuth implements FirebaseServiceInterface {
...
+
+  /**
+   * Looks up the tenant identified by the provided tenant ID and returns a promise that is
+   * fulfilled with the corresponding tenant if it is found.
+   *
+   * @param {string} tenantId The tenant ID of the tenant to look up.
+   * @return {Promise<Tenant>} A promise that resolves with the corresponding tenant.
+   */
+  public getTenant(tenantId: string): Promise<Tenant> {
+    return (this.authRequestHandler as AuthRequestHandler).getTenant(tenantId)
```
Comment (hiranya911) [inline src/auth/auth.ts]: You can use generics and get rid of these casts if you want to.

```
abstract class AbstractRequestHandler { }

class AuthRequestHandler extends AbstractRequestHandler {
  public getTenant() { }
}

class TenantAwareRequestHandler extends AbstractRequestHandler { }

abstract class BaseAuth<T extends AbstractRequestHandler> {
  protected requestHandler: T;
}

class Auth extends BaseAuth<AuthRequestHandler> {
  public getTenant() {
    // requestHandler is treated as AuthRequestHandler by the compiler
    this.requestHandler.getTenant();
  }
}
```

Comment (bojeil-google) [reply src/auth/auth.ts]: Good idea. Done.

File: test/unit/auth/auth.spec.ts [hunk 2]
```diff
@@ -3052,6 +3053,506 @@ AUTH_CONFIGS.forEach((testConfig) => {
...
+        };
+        const emptyListTenantsResponse: any = {
+          tenants: [],
+        };
+        const emptyExpectedResult: any = {
+          tenants: [],
+        };
+        // Stubs used to simulate underlying API calls.
+        let stubs: sinon.SinonStub[] = [];
+        afterEach(() => {
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Nit: Add an empty line before this for clarity.

File: test/unit/auth/auth.spec.ts [hunk 3]
```diff
@@ -3052,6 +3053,506 @@ AUTH_CONFIGS.forEach((testConfig) => {
...
+          stubs.push(listTenantsStub);
+          return (auth as Auth).listTenants()
+            .then((response) => {
+              expect(response).to.deep.equal(expectedResult);
+              // Confirm underlying API called with expected parameters.
+              expect(listTenantsStub)
+                .to.have.been.calledOnce.and.calledWith(undefined, undefined);
+            });
+        });
+
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Redundant empty line

File: test/unit/auth/auth.spec.ts [hunk 4]
```diff
@@ -3052,6 +3053,506 @@ AUTH_CONFIGS.forEach((testConfig) => {
...
+            .then(() => {
+              throw new Error('Unexpected success');
+            })
+            .catch((error) => {
+              expect(error).to.have.property('code', 'auth/argument-error');
+            });
+        });
+
+        it('should be rejected given TenantOptions with invalid property', () => {
+          return (auth as Auth).updateTenant(tenantId, {type: 'lightweight'})
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Can we make it clear what the invalid property is? Is it the key or the value that makes it invalid? Depending on the intention of the test either `{invalid: 'lightweight'}` or `{type: 'invalid'}`. If neither is appropriate, update the description of the test.

File: test/unit/auth/auth.spec.ts (hunk 2 above)
Comment (bojeil-google) [reply test/unit/auth/auth.spec.ts]: done

File: test/unit/auth/auth.spec.ts (hunk 3 above)
Comment (bojeil-google) [reply test/unit/auth/auth.spec.ts]: done

File: test/unit/auth/auth.spec.ts (hunk 4 above)
Comment (bojeil-google) [reply test/unit/auth/auth.spec.ts]: I updated test description and added a comment here and in the `createTenant` test similar to this.

**Pull request merged**